x-pack/plugins/osquery/server/_telemetry/sender.ts

```diff
@@ -0,0 +1,193 @@
+
+  constructor(logger: Logger) {
+    this.logger = logger;
+  }
+
+  public setup(telemetrySetup?: TelemetryPluginSetup) {
+    this.telemetrySetup = telemetrySetup;
+  }
+
+  public async start(telemetryStart?: TelemetryPluginStart, core?: CoreStart) {
+    this.telemetryStart = telemetryStart;
+    this.esClient = core?.elasticsearch.client.asInternalUser;
+    this.clusterInfo = await this.fetchClusterInfo();
+
+    this.logger.debug(`Starting local task`);
+    setTimeout(() => {
+      this.sendIfDue();
+      this.intervalId = setInterval(() => this.sendIfDue(), this.checkIntervalMs);
+    }, this.initialCheckDelayMs);
+  }
+
+  public stop() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  public queueTelemetryEvents<T extends OsqueryTelemetryChannel>(
+    channel: T,
+    events: Array<OsqueryTelemetryChannelEvents[T]>
+  ) {
+    if (!this.queuesPerChannel[channel]) {
+      this.queuesPerChannel[channel] = new TelemetryQueue<OsqueryTelemetryChannelEvents[T]>();
+    }
+    this.queuesPerChannel[channel].addEvents(cloneDeep(events));
+  }
+
+  public async isTelemetryOptedIn() {
+    this.isOptedIn = await this.telemetryStart?.getIsOptedIn();
+    return this.isOptedIn === true;
+  }
+
+  private async sendIfDue() {
+    if (this.isSending) {
+      return;
+    }
+
+    try {
+      this.isSending = true;
+
+      this.isOptedIn = await this.isTelemetryOptedIn();
+      if (!this.isOptedIn) {
+        this.logger.debug(`Telemetry is not opted-in.`);
+        for (const channel of Object.keys(this.queuesPerChannel)) {
+          this.queuesPerChannel[channel].clearEvents();
+        }
+        this.isSending = false;
+        return;
+      }
+
+      for (const channel of Object.keys(this.queuesPerChannel)) {
+        await this.sendEvents(
+          await this.fetchTelemetryUrl(channel),
+          this.clusterInfo,
+          this.queuesPerChannel[channel]
+        );
+      }
+    } catch (err) {
+      this.logger.warn(`Error sending telemetry events data: ${err}`);
+    }
+
+    this.isSending = false;
+  }
+
+  private async fetchClusterInfo(): Promise<InfoResponse> {
+    if (this.esClient === undefined || this.esClient === null) {
+      throw Error('elasticsearch client is unavailable: cannot retrieve cluster information');
+    }
+
+    const { body } = await this.esClient.info();
+    return body;
+  }
+
+  public async sendEvents(
+    telemetryUrl: string,
+    clusterInfo: InfoResponse | undefined,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    queue: TelemetryQueue<any>
+  ) {
+    const events = queue.getEvents();
+    if (events.length === 0) {
+      return;
+    }
+
+    try {
+      this.logger.debug(`Telemetry URL: ${telemetryUrl}`);
+
+      queue.clearEvents();
+
+      this.logger.debug(JSON.stringify(events));
+
+      await this.send(
+        events,
+        telemetryUrl,
+        clusterInfo?.cluster_uuid,
```

Comment:
Please also send `cluster_name` as a header. We have updated our instrumentation on both sides of the pipeline